### ISSUE TITLE
[REQUEST] FP32 gradient accumulation during BF16 training in ZeRO2/ZeRO3

### ISSUE THREAD

**Issue by michaelroyzen** [microsoft/DeepSpeed#2768]
**Is your feature request related to a problem? Please describe.**
Training occasionally diverges when I train in BF16 with ZeRO enabled because accumulation is also done in BF16 and not FP32. Interestingly, FP32 accumulation is supported for BF16 training in DeepSpeed -- but only when ZeRO is disabled. Why is this? It doesn't make any sense to me.

**Describe the solution you'd like**
BF16 training should always accumulate gradients in FP32, regardless of ZeRO mode.

Linking: #2742 

@stas00 @tjruwase

**Comment by stas00** [comment]
Yes! I'm glad you are adding your voice. My team needs this too.

For fp16 grad accum in fp16 seems to work just fine. But clearly not for bf16.

I also proposed to Tunji to optionally explore implementing Kahan summation which should overcome this with bf16 accumulation. But it'd also need an additional bf16 buffer. So memory-wise Kahan summation should be the same as using an FP32 accumulator, but speed-wise I think it might be a bit faster. Need to benchmark.

**Comment by tjruwase** [comment]
@michaelroyzen, @stas00, thanks for raising this. We are working on it.

**Comment by tjruwase** [comment]
@michaelroyzen, closing as duplicate of #2352. Rest assured this is high priority.
